### ISSUE TITLE
twilio sms HTTP retrieval failure

### ISSUE THREAD

**Issue by MikeLeonard** [coddingtonbear/inthe.am#173]
Hi  Love Inthe.AM so much it's a real pleasure to use the way it's designed, keep up the good work!

I'm just having one issue with the sms feature....

I've just paid for a twilio number and followed the configuration instructions (ie pasted in my auth key, left the whitelist numbers box as blank, copied https://inthe.am/api/v1/task/my_user_name/sms to the POST htpp box on the twilio page for my number.  

I'm having an issue however... nothing arrives to taskwarrior/Inthe.AM.  Checking the twilio page for my new number, it reports the error '11200 HTTP retrieval failure' when I click on one of the texts I sent through to test.

The description of the error is:  'An attempt to retrieve content from https://inthe.am/api/v1/task/my_user_name/sms returned the HTTP status code 500.'

I've pasted below the exact description of the error from the twilio site.

I'm not an expert in these things I'm afraid but am keen to get this working (always travelling, texting to taskwarrior would be amazing!).  Any chance there's something obvious I've missed or a way you are aware of that I can fix this?

Happy to test something or provide more info if that would help

Really would appreciate any help

Thanks :)

Mike

............................
here was a failure attempting to retrieve the contents of this URL.

An 11200 error is an indicator of a connection failure between Twilio and your service. When Twilio requests a page from your server, we wait a maximum of 15 seconds for a response. A connection failure will occur if no response is returned in that time.

There are many reasons a connection timeout can occur; common causes are long running database queries or outside processes and calls to external systems taking a long time to return. It may be possible your application experienced one of these issues. If you are encountering this error only intermittently, it is possible that your web server was temporarily unavailable or experiencing a network outage.

If your debugger is reporting a 502 Bad Gateway error, this may mean that Twilio's internal server had trouble retrieving content from your website. Your request must contain a Content-Type that is valid. Twilio may also have had problems resolving your DNS name to an IP address, or issues with the network connection. Check that your web server is started, and is available over the public Internet.

Web server returned a 4xx or 5xx HTTP response to Twilio
Misconfigured Web Server
Network disruptions between Twilio and your web server
No Content-Type header attached to response
Content-Type doesn't match actual content, e.g. an MP3 file that is being served with Content-Type: audio/x-wav, instead of Content-Type: audio/mpeg
Double check that your TwiML URL does not return a 4xx or 5xx error
Make certain that the URL does not perform a 302 redirect to an invalid URL
Confirm the URL requested is not protected by HTTP Auth
Make sure your web server allows HTTP POST requests to static resources (if the URL refers to .xml or .html files)
Verify your web server is up and responsive
Check to see that the URL host is not a private or local IP address
Verify the ping times and packet loss between your web server and www.twilio.com

**Comment by MikeLeonard** [comment]
Ah ok more info... I've just noticed this in the Inthe.AM log (well, first I noticed that there's a lot!  Then I noticed this :) )

Incoming SMS for emailmleonard, but error encountered while attempting to build request validator: 'HTTP_X_TWILIO_SIGNATURE'.

Is that helpful?

Thanks again

**Comment by coddingtonbear** [comment]
Oh, hey there -- I was actually just about to send you an e-mail!

I see the error [appearing in the log](http://sentry.adamcoddington.net/the-proletariat/in-the-am/group/852/) (and you probably see an abbreviated one if you click the 'Activity Log' button while logged-in to Inthe.AM), and looking through the source, it looks like this error message will appear if the incoming request doesn't include the authentication header we're looking for -- I'm a little surprised that such a message might appear if the request did originate with Twilio, but I've just increased the logging level a little bit (via 81330eb807b7b2742223e8df98862e582b764b10, just released a moment ago) to see if I can help you troubleshoot.

Could you try sending-in another SMS via Twilio so I have a little more logging information to work with?

**Comment by MikeLeonard** [comment]
Blimey talk about a quick reply :)   Thanks.

I've just texted 'add more testing to generate log' to my twilio number.  I
see exactly the same as before on the twilio site.. and at the moment
there's nothing in the log on Inthe.AM, should something appear in the log
soon do you think?

**Comment by coddingtonbear** [comment]
Oh my -- I see what the problem is.  Just a moment, let me push up a patch.

**Comment by coddingtonbear** [comment]
How embarrassing -- thanks for catching that!  Earlier this week, I made several changes to clean up some early cruft, and as part of that removed what at the time looked like a one-off method of fetching the user's taskstore; unfortunately, in this specific case, the one-off method of selecting the taskstore is necessary since the request isn't actually authenticated until after the store is gathered.

TLDR; fixed via aa177ad0023640809ea1b55b601a21fbb35bc691; terribly embarassed; sorry about that!

Give it a shot now -- I just tested it out using my twilio account, and things appear to be working just fine.

**Comment by MikeLeonard** [comment]
Hi

Yep I can confirm that it's working now.  Funny that I've been meaning to
set up sms for a while now and manage to do it in the week where a code
change affected it.  Glad that I could help you discover and fix that -
thanks again for the very quick response :)

**Comment by coddingtonbear** [comment]
No problem!

**Comment by kiranbakale** [comment]
hey

**Comment by kiranbakale** [comment]
https://github.com/elyserobinson/whatsappstockbot/issues/1#issue-947000083

**please can you go through this files **
